physical/raft/raft.go

```diff
@@ -395,33 +400,71 @@ func NewRaftBackend(conf map[string]string, logger log.Logger) (physical.Backend
 		snap = raft.NewInmemSnapshotStore()
 	} else {
 		// Create the base raft path.
-		path := filepath.Join(path, raftState)
-		if err := EnsurePath(path, true); err != nil {
+		raftBasePath := filepath.Join(path, raftState)
+		if err := EnsurePath(raftBasePath, true); err != nil {
 			return nil, err
 		}
+		dbPath := filepath.Join(raftBasePath, "raft.db")
 
-		// Create the backend raft store for logs and stable storage.
-		dbPath := filepath.Join(path, "raft.db")
-		opts := boltOptions(dbPath)
-		raftOptions := raftboltdb.Options{
-			Path:        dbPath,
-			BoltOptions: opts,
+		// Check if they're opting in to raft-wal storage
+		useRaftWal := false
+		if walRaw, ok := conf[raftWalConfigKey]; ok {
+			useRaftWal, err = strconv.ParseBool(walRaw)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse %s config value %q as a boolean: %w", raftWalConfigKey, walRaw, err)
+			}
 		}
-		store, err := raftboltdb.New(raftOptions)
-		if err != nil {
-			return nil, err
+
+		// If the existing raft db exists from a previous use of BoltDB, warn about this and continue to use BoltDB
+		_, err := os.Stat(dbPath)
+		if useRaftWal && errors.Is(err, fs.ErrExist) {
+			logger.Warn("raft is configured to use raft-wal for storage but existing raft.db detected. raft-wal config will be ignored.")
+			useRaftWal = false
+		}
+
+		if useRaftWal {
+			raftWalPath := filepath.Join(raftBasePath, raftWalDir)
+			if err := EnsurePath(raftWalPath, true); err != nil {
+				return nil, err
+			}
+
+			// TODO: is this the right way to do metrics? Do they need to be connected to the core sink somehow?
+			mc := walmetrics.NewGoMetricsCollector([]string{"raft", "wal"}, nil, nil)
+
+			// TODO: do we need to expose any tunables here in the config?
```

Comment:
Equivalent in Consul: https://github.com/hashicorp/consul/blob/e235c8be3c67ed1389af017a76b29a8452b86453/agent/consul/server.go#L1032-L1050

Yes this is right for metrics - go-metrics (for better or worse) uses a global instance against which all sinks are registered so no dependency injection needed here. The prefix is probably OK too but we should manually check this as part of testing before we merge that the metrics it emits look reasonable and have labels consistent with other parts of Vault.

The only tunable I exposes in Consul was segment size. I don't even know if that is really important to be honest. About the only reason I can think you'd want to tune that is if you were running on absolutely tiny hardware like an embedded device where having at least two 64MiB segment files would be too costly and throughput is super low anyway. It's possible someone one day will have write throughput so high that segments larger than 64MiB are important either because of rotation overhead or because then end up with so many segments to retain the logs they need that there is extra metadata overhead. But honestly I suspect we are at least one order of magnitude of write scale away from that being an issue in Vault right now even at our biggest.

I tend to prefer fewer tunables since then you have to provide guidance on when to change them which is always a subtle discussion but one argument might be to include it for consistency with Consul?

Reply:
I always prefer fewer tunables too, because my experience is if there's a tunable available, people will tune it, even if they don't need to, and usually to their own detriment. I might skip all tunables for now, even segment size, just in the name of simplicity. It's always easier to add tunables later if we need to, whether for consistency with Consul or another reason, but it's hard to remove them once they're there.